sklearn/model_selection/_split.py

```diff
@@ -2506,6 +2508,155 @@ def _build_repr(self):
+    ...     print("TRAIN:", train_idx, "TEST:", test_idx)
+    ...     print(
+    ...         "TRAIN GROUP:", groups[train_idx], "TEST GROUP:", groups[test_idx]
+    ...     )
+    TRAIN: [0 1 2 3] TEST: [4 5 6]
+    TRAIN GROUP: ['a' 'a' 'a' 'a'] TEST GROUP: ['b' 'b' 'b']
+    TRAIN: [0 1 2 3 4 5 6] TEST: [ 7  8  9 10]
+    TRAIN GROUP: ['a' 'a' 'a' 'a' 'b' 'b' 'b'] TEST GROUP: ['c' 'c' 'c' 'c']
+    TRAIN: [ 0  1  2  3  4  5  6  7  8  9 10] TEST: [11 12 13]
+    TRAIN GROUP: ['a' 'a' 'a' 'a' 'b' 'b' 'b' 'c' 'c' 'c' 'c'] TEST GROUP: ['d' 'd' 'd']
+    """
+
+    @_deprecate_positional_args
+    def __init__(self, n_splits=5, *, max_train_size=None, test_size=None, gap=0):
+        super().__init__(n_splits, shuffle=False, random_state=None)
+        self.max_train_size = max_train_size
+        self.test_size = test_size
+        self.gap = gap
+
+    def split(self, X, y=None, groups=None):
+        """Generate indices to split data into training and test set.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like of shape (n_samples,)
+            Always ignored, exists for compatibility.
+
+        groups : array-like of shape (n_samples,)
+            Group labels for the samples used while splitting the dataset into
+            train/test set.
+
+        Yields
+        ------
+        train : ndarray
+            The training set indices for that split.
+
+        test : ndarray
+            The testing set indices for that split.
+        """
+        if groups is None:
+            raise ValueError("The 'groups' parameter should not be None")
+        X, y, groups = indexable(X, y, groups)
+        n_samples = _num_samples(X)
+        n_splits = self.n_splits
+        n_folds = n_splits + 1
+        group_dict = {}
+        u, ind = np.unique(groups, return_index=True)
+        unique_groups = u[np.argsort(ind)]
+        n_samples = _num_samples(X)
+        n_groups = _num_samples(unique_groups)
+        for idx in np.arange(n_samples):
+            if groups[idx] in group_dict:
+                if idx - group_dict[groups[idx]][-1] == 1:
+                    group_dict[groups[idx]].append(idx)
+                else:
+                    raise ValueError(
+                        (
+                            "The groups should be continuous."
+                            " Found a non-countinuous group at"
+                            " index={0}"
+                        ).format(idx)
+                    )
+            else:
+                group_dict[groups[idx]] = [idx]
+            if n_folds > n_groups:
+                raise ValueError(
+                    (
+                        "Cannot have number of folds={0} greater than"
+                        " the number of groups={1}"
+                    ).format(n_folds, n_groups)
```

Comment:
I think that we can speed up this dictionary generation with NumPy:

```python
reordered_unique_groups, indices = np.unique(groups, return_index=True)
original_order_group = indices.argsort()
groups_indices = np.split(np.arange(groups.size), indices[original_order_group][1:])
group_dict = {
   g: i for g, i in zip(reordered_unique_groups[original_order_group], groups_indices)
}
```

We can then make a check to be sure that the indices are consistent.